File: hn2016_falwa/f90_modules/matrix_b4_inversion.f90
SUBROUTINE matrix_b4_inversion(k,jmax,kmax,nd,jb,jd,z,statn,qref,ckref,&
        a, om, dz, h, rr, cp, &
        qjj,djj,cjj,rj,tj,sjk,tjk)

  integer, INTENT(in) :: k, jmax, kmax, nd, jb, jd
  REAL, INTENT(in) :: z(kmax),statn(kmax),qref(nd,kmax),ckref(nd,kmax)
  REAL, INTENT(in) :: a, om, dz, h, rr, cp
  REAL, INTENT(OUT) :: qjj(jd-2,jd-2),djj(jd-2,jd-2),cjj(jd-2,jd-2),rj(jd-2),tj(jd-2)
  REAL, INTENT(INOUT) :: sjk(jd-2,jd-2,kmax-1),tjk(jd-2,kmax-1)
  REAL :: xjj(jd-2,jd-2), u(jd,kmax)
  REAL :: sjj(jd-2,jd-2)

  rkappa = rr/cp
  pi = acos(-1.)
  dp = pi/float(jmax-1)

  zp = 0.5*(z(k+1)+z(k))
  zm = 0.5*(z(k-1)+z(k))
  statp = 0.5*(statn(k+1)+statn(k))
  statm = 0.5*(statn(k-1)+statn(k))
  cjj(:,:) = 0.
  djj(:,:) = 0.
  qjj(:,:) = 0.
  sjj(:,:) = sjk(:,:,k)
  tj(:) = tjk(:,k)
  do jj = jb+2,(nd-1)
    j = jj - jb
    phi0 = float(jj-1)*dp
    phip = (float(jj)-0.5)*dp
    phim = (float(jj)-1.5)*dp
    cos0 = cos(phi0)
    cosp = cos(phip)
    cosm = cos(phim)
    sin0 = sin(phi0)
    sinp = sin(phip)
    sinm = sin(phim)

    fact = 4.*om*om*h*a*a*sin0*dp*dp/(dz*dz*rr*cos0)
    amp = exp(-zp/h)*exp(rkappa*zp/h)/statp
    amp = amp*fact*exp(z(k)/h)
    amm = exp(-zm/h)*exp(rkappa*zm/h)/statm
    amm = amm*fact*exp(z(k)/h)

    ! ***** Specify A, B, C, D, E, F (Eqs. 4-9) *****
    ajk = 1./(sinp*cosp)
    bjk = 1./(sinm*cosm)
    cjk = amp
    djk = amm
    ejk = ajk+bjk+cjk+djk
    fjk = -0.5*a*dp*(qref(jj+1,k)-qref(jj-1,k))

    ! ***** Specify rk (Eq. 15) ****

    ! **** North-south boundary conditions ****
    u(jd,k) = 0.
    phi0 = dp*float(jb)
    !        u(1,k) = ubar(jb+1,k)*cos(phi0)
    u(1,k) = ckref(jb+1,k)/(2.*pi*a)-om*a*cos(phi0)

    rj(j-1) = fjk
    if(j.eq.2) rj(j-1) = fjk - bjk*u(1,k)
    if(j.eq.jd-1) rj(j-1) = fjk - ajk*u(jd,k)

    ! ***** Specify Ck & Dk (Eqs. 18-19) *****
    cjj(j-1,j-1) = cjk
    djj(j-1,j-1) = djk

    ! **** Specify Qk (Eq. 17) *****
    qjj(j-1,j-1) = -ejk
    if(j-1.ge.1.and.j-1.lt.jd-2) then
      qjj(j-1,j) = ajk
    endif
    if(j-1.gt.1.and.j-1.le.jd-2) then
      qjj(j-1,j-2) = bjk
    endif
  enddo

  ! **** Compute Qk + Ck Sk *******
  do i = 1,jd-2
    do j = 1,jd-2
      xjj(i,j) = 0.
      do kk = 1,jd-2
        xjj(i,j) = xjj(i,j)+cjj(i,kk)*sjj(kk,j)
      enddo
      qjj(i,j) = qjj(i,j)+xjj(i,j)
    enddo
  enddo

END SUBROUTINE matrix_b4_inversion

File: src/falwa/numba_modules/matrix_b4_inversion.py
"""
Numba implementation of matrix_b4_inversion.

This module provides a JIT-compiled Python implementation of the Fortran
subroutine `matrix_b4_inversion` for setting up matrices before inversion
in the direct solver algorithm.

.. versionadded:: 2.4.0

Notes
-----
Arrays use C-order indexing:
- 2D lat-height arrays: [k, j]
- 3D arrays: [k, j1, j2]
"""

import numpy as np
from numba import njit
from typing import Tuple


@njit(cache=True)
def _matrix_b4_inversion_core(
    k: int,
    jmax: int,
    kmax: int,
    nd: int,
    jb: int,
    jd: int,
    z: np.ndarray,
    statn: np.ndarray,
    qref: np.ndarray,
    ckref: np.ndarray,
    sjk: np.ndarray,
    a: float,
    om: float,
    dz: float,
    h: float,
    rr: float,
    cp: float
) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """
    Core implementation of matrix_b4_inversion.
    """
    rkappa = rr / cp
    pi = np.arccos(-1.0)
    dp = pi / float(jmax - 1)
    
    zp = 0.5 * (z[k] + z[k - 1])
    zm = 0.5 * (z[k - 2] + z[k - 1])
    statp = 0.5 * (statn[k] + statn[k - 1])
    statm = 0.5 * (statn[k - 2] + statn[k - 1])
    
    # Initialize output arrays
    cjj = np.zeros((jd - 2, jd - 2), dtype=np.float64)
    djj = np.zeros((jd - 2, jd - 2), dtype=np.float64)
    qjj = np.zeros((jd - 2, jd - 2), dtype=np.float64)
    rj = np.zeros(jd - 2, dtype=np.float64)
    
    # Copy sjk slice (C-order: sjk[k, j1, j2])
    sjj = np.zeros((jd - 2, jd - 2), dtype=np.float64)
    for i in range(jd - 2):
        for j in range(jd - 2):
            sjj[i, j] = sjk[k - 1, i, j]
    
    # Initialize u array for boundary conditions (C-order: [k, j])
    u = np.zeros((kmax, jd), dtype=np.float64)
    
    # Fortran: do jj = jb+2, (nd-1)
    for jj in range(jb + 2, nd):
        j = jj - jb
        phi0 = float(jj - 1) * dp
        phip = (float(jj) - 0.5) * dp
        phim = (float(jj) - 1.5) * dp
        cos0 = np.cos(phi0)
        cosp = np.cos(phip)
        cosm = np.cos(phim)
        sin0 = np.sin(phi0)
        sinp = np.sin(phip)
        sinm = np.sin(phim)
        
        fact = 4.0 * om * om * h * a * a * sin0 * dp * dp / (dz * dz * rr * cos0)
        amp = np.exp(-zp / h) * np.exp(rkappa * zp / h) / statp
        amp = amp * fact * np.exp(z[k - 1] / h)
        amm = np.exp(-zm / h) * np.exp(rkappa * zm / h) / statm
        amm = amm * fact * np.exp(z[k - 1] / h)
        
        # Specify A, B, C, D, E, F (Eqs. 4-9)
        ajk = 1.0 / (sinp * cosp)
        bjk = 1.0 / (sinm * cosm)
        cjk = amp
        djk = amm
        ejk = ajk + bjk + cjk + djk
        fjk = -0.5 * a * dp * (qref[k - 1, jj] - qref[k - 1, jj - 2])  # C-order: qref[k, j]
        
        # North-south boundary conditions
        u[k - 1, jd - 1] = 0.0
        phi0_bc = dp * float(jb)
        u[k - 1, 0] = ckref[k - 1, jb] / (2.0 * pi * a) - om * a * np.cos(phi0_bc)
        
        rj[j - 2] = fjk
        if j == 2:
            rj[j - 2] = fjk - bjk * u[k - 1, 0]
        if j == jd - 1:
            rj[j - 2] = fjk - ajk * u[k - 1, jd - 1]
        
        # Specify Ck & Dk (Eqs. 18-19)
        cjj[j - 2, j - 2] = cjk
        djj[j - 2, j - 2] = djk
        
        # Specify Qk (Eq. 17)
        qjj[j - 2, j - 2] = -ejk
        if j - 1 >= 1 and j - 1 < jd - 2:
            qjj[j - 2, j - 1] = ajk
        if j - 1 > 1 and j - 1 <= jd - 2:
            qjj[j - 2, j - 3] = bjk
    
    # Compute Qk + Ck Sk
    xjj = np.zeros((jd - 2, jd - 2), dtype=np.float64)
    for i in range(jd - 2):
        for j in range(jd - 2):
            xjj[i, j] = 0.0
            for kk in range(jd - 2):
                xjj[i, j] += cjj[i, kk] * sjj[kk, j]
            qjj[i, j] = qjj[i, j] + xjj[i, j]
    
    return qjj, djj, cjj, rj


def matrix_b4_inversion(
    k: int,
    jmax: int,
    jb: int,
    jd: int,
    z: np.ndarray,
    statn: np.ndarray,
    qref: np.ndarray,
    ckref: np.ndarray,
    sjk: np.ndarray,
    a: float,
    om: float,
    dz: float,
    h: float,
    rr: float,
    cp: float
) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """
    Set up matrices before inversion in the direct solver algorithm.
    
    Parameters
    ----------
    k : int
        Current vertical level index (1-indexed as in Fortran).
    jmax : int
        Number of latitude points.
    jb : int
        Lower bounding latitude index.
    jd : int
        nd - lower bounding latitude.
    z : np.ndarray
        Height levels, shape (kmax,).
    statn : np.ndarray
        Static stability, shape (kmax,).
    qref : np.ndarray
        Reference QGPV, shape (kmax, nd).
    ckref : np.ndarray
        Reference Kelvin circulation, shape (kmax, nd).
    sjk : np.ndarray
        S matrix from previous iteration, shape (kmax-1, jd-2, jd-2).
    a : float
        Earth radius.
    om : float
        Earth rotation rate.
    dz : float
        Vertical grid spacing.
    h : float
        Scale height.
    rr : float
        Gas constant.
    cp : float
        Specific heat.
        
    Returns
    -------
    qjj : np.ndarray
        Q matrix, shape (jd-2, jd-2).
    djj : np.ndarray
        D matrix, shape (jd-2, jd-2).
    cjj : np.ndarray
        C matrix, shape (jd-2, jd-2).
    rj : np.ndarray
        R vector, shape (jd-2,).
    """
    kmax, nd = qref.shape
    
    return _matrix_b4_inversion_core(
        int(k), int(jmax), int(kmax), int(nd), int(jb), int(jd),
        z, statn, qref, ckref, sjk,
        float(a), float(om), float(dz), float(h), float(rr), float(cp)
    )
